pkg/machinery/client/events.go

```diff
@@ -123,6 +123,86 @@ func (c *Client) EventsWatch(ctx context.Context, watchFunc func(<-chan Event),
 	}
 }
 
+// EventResult is the result of an event watch, containing either an Event or an error.
+type EventResult struct {
+	// Event is the event that was received.
+	Event Event
+	// Err is the error that occurred.
+	Error error
+}
+
+// EventsWatchV2 watches events of a single node and wraps the Events by providing a simpler interface.
+// It blocks until the first (empty) event is received, then spawns a goroutine that sends events to the given channel.
+// EventResult objects sent into the channel contain either the errors or the received events.
+//
+//nolint:gocyclo
+func (c *Client) EventsWatchV2(ctx context.Context, ch chan<- EventResult, opts ...EventsOptionFunc) error {
+	stream, err := c.Events(ctx, opts...)
+	if err != nil {
+		return fmt.Errorf("error fetching events: %w", err)
+	}
+
+	if err = stream.CloseSend(); err != nil {
+		return err
+	}
+
+	defaultNode := RemotePeer(stream.Context())
+
+	// receive first (empty) watch event
+	_, err = stream.Recv()
+	if err != nil {
+		if err == io.EOF || StatusCode(err) == codes.Canceled {
+			return nil
+		}
+
+		return fmt.Errorf("error while watching events: %w", err)
+	}
+
+	go func() {
+		for {
+			event, err := stream.Recv()
+			if err != nil {
+				if err == io.EOF || StatusCode(err) == codes.Canceled {
+					return
+				}
+
+				ch <- EventResult{Error: fmt.Errorf("error while watching events: %w", err)}
```

Comment:
I would probably wrap all `ch <-` sends with `select` on `ctx.Done()` to be on the safe side - if the caller walked away and cancelled the context, there's not much we can do, it might not be consuming the events anymore